xkafka/producer.go

```diff
@@ -0,0 +1,183 @@
+package xkafka
+
+import (
+	"context"
+	"strings"
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/pkg/errors"
+)
+
+// Producer manages the production of messages to kafka topics.
+// It provides both synchronous and asynchronous publish methods
+// and a channel to stream delivery events.
+type Producer struct {
+	config              options
+	kafka               producerClient
+	events              chan kafka.Event
+	middlewares         []middleware
+	wrappedPublish      Handler
+	wrappedAsyncPublish Handler
+}
+
+// NewProducer creates a new Producer.
+func NewProducer(name string, opts ...Option) (*Producer, error) {
+	cfg := defaultProducerOptions()
+
+	for _, opt := range opts {
+		opt.apply(&cfg)
+	}
+
+	_ = cfg.configMap.SetKey("bootstrap.servers", strings.Join(cfg.brokers, ","))
+	_ = cfg.configMap.SetKey("client.id", name)
+
+	producer, err := cfg.producerFn(&cfg.configMap)
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Producer{
+		config: cfg,
+		kafka:  producer,
+		events: producer.Events(),
+	}
+
+	p.wrappedPublish = p.publish()
+	p.wrappedAsyncPublish = p.asyncPublish()
+
+	return p, nil
+}
+
+// Use appends a MiddlewareFunc to the chain.
+// Middleware can be used to intercept or otherwise modify, process or skip messages.
+// They are executed in the order that they are applied to the Producer.
+func (p *Producer) Use(mwf ...MiddlewareFunc) {
```

Comment:
This function also needs to be tested, can you add table tests where the order of the middlewares is also tested.